kernel/generic/src/mm/as.c

```diff
@@ -313,76 +377,39 @@ NO_TRACE static bool check_area_conflicts(as_t *as, uintptr_t addr,
 		return false;
 
 	/*
-	 * The leaf node is found in O(log n), where n is proportional to
-	 * the number of address space areas belonging to as.
-	 * The check for conflicts is then attempted on the rightmost
-	 * record in the left neighbour, the leftmost record in the right
-	 * neighbour and all records in the leaf node itself.
+	 * To determine if we overlap with another area, we just need
+	 * to look at overlap with the last area with base address <=
+	 * to ours and on the first area with base address > than ours.
+	 *
+	 * First find last area with <= base address.
 	 */
-	btree_node_t *leaf;
-	as_area_t *area =
-	    (as_area_t *) btree_search(&as->as_area_btree, addr, &leaf);
-	if (area) {
-		if (area != avoid)
-			return false;
-	}
-
-	/* First, check the two border cases. */
-	btree_node_t *node =
-	    btree_leaf_node_left_neighbour(&as->as_area_btree, leaf);
-	if (node) {
-		area = (as_area_t *) node->value[node->keys - 1];
+	odlink_t *odlink = odict_find_leq(&as->as_areas, &addr, NULL);
+	if (odlink != NULL) {
+		as_area_t *area = odict_get_instance(odlink, as_area_t,
+		    las_areas);
 
 		if (area != avoid) {
 			mutex_lock(&area->lock);
-
-			/*
-			 * If at least one of the two areas are protected
-			 * by the AS_AREA_GUARD flag then we must be sure
-			 * that they are separated by at least one unmapped
-			 * page.
-			 */
-			int const gp = (guarded ||
-			    (area->flags & AS_AREA_GUARD)) ? 1 : 0;
-
-			/*
-			 * The area comes from the left neighbour node, which
-			 * means that there already are some areas in the leaf
-			 * node, which in turn means that adding gp is safe and
-			 * will not cause an integer overflow.
-			 */
-			if (overlaps(addr, P2SZ(count), area->base,
-			    P2SZ(area->pages + gp))) {
+			if (area_is_conflicting(addr, count, guarded, area)) {
 				mutex_unlock(&area->lock);
 				return false;
 			}
 
 			mutex_unlock(&area->lock);
 		}
+
+		/* Next area */
+		odlink = odict_next(odlink, &as->as_areas);
 	}
 
-	node = btree_leaf_node_right_neighbour(&as->as_area_btree, leaf);
-	if (node) {
-		area = (as_area_t *) node->value[0];
+	/* Next area, if any, is the first with base > than our base address */
+	if (odlink != NULL) {
```

Comment:
How does hit work if there was no area with base <= our base address, but there is an area with base > our base address? It seems to me odlink will be NULL and so the branch for the latter will not get executed.

Reply:
Nice catch, fixed.